src/main/java/org/spongepowered/api/world/teleport/TeleportHelperKernel.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.world.teleport;
+
+import com.flowpowered.math.vector.Vector3i;
+import org.spongepowered.api.CatalogType;
+import org.spongepowered.api.block.BlockState;
+import org.spongepowered.api.util.Tristate;
+import org.spongepowered.api.util.annotation.CatalogedBy;
+import org.spongepowered.api.world.Location;
+import org.spongepowered.api.world.TeleportHelper;
+import org.spongepowered.api.world.World;
+
+/**
+ * A {@link TeleportHelperKernel} contains routines to determine
+ * whether a block is a suitable floor or body material.
+ */
+@CatalogedBy(TeleportHelperKernels.class)
+public interface TeleportHelperKernel extends CatalogType {
+
+    /**
+     * Tests whether the location in question is valid, regardless of whether
+     * the block is safe or not.
+     *
+     * <p>Returning {@link Tristate#TRUE} marks the location as valid and will
+     * causes the parent {@link TeleportHelper} to return this location,
+     * returning {@link Tristate#FALSE} marks the location as invalid and
+     * causes the parent {@link TeleportHelper} to move onto the next block to
+     * check.</p>
+     *
+     * @param world The {@link World} to check.
+     * @param position The {@link Vector3i} (block position) to check.
+     * @return A {@link Tristate}.
+     */
+    Tristate isValidLocation(World world, Vector3i position);
+
+    /**
+     * Tests whether a {@link BlockState} should be considered a safe block
+     * to land on.
+     *
+     * @param blockState The {@link BlockState} to check
+     * @return <code>true</code> if the material should be safe to land on.
+     */
+    boolean isSafeFloorMaterial(BlockState blockState);
```

Comment:
Ah OK. I see where you're coming from, but I don't know what a kernel would suggest is unknown, and in the case of one kernel, then there needs to be a resolution. I'd think it's up to the wrapping kernel to decide what to do, so in the case of a modded block kernel that works with our `DEFAULT`, I would expect it would bypass our kernel and just use it's own checks, or with a kernel that adds extra restrictions to just `AND` the checks, similarly, one that loosens protections to `OR` or selective bypass the wrapped checks.